eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/BbodySpec.hs

```diff
@@ -0,0 +1,139 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE NumericUnderscores #-}
+{-# LANGUAGE OverloadedLists #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Test.Cardano.Ledger.Conway.Imp.BbodySpec (
+  spec,
+) where
+
+import Cardano.Ledger.BHeaderView (BHeaderView (..))
+import Cardano.Ledger.Babbage.Core
+import Cardano.Ledger.BaseTypes (BlocksMade (..), ProtVer (..), SlotNo (..))
+import Cardano.Ledger.Block
+import Cardano.Ledger.Coin (Coin (..))
+import Cardano.Ledger.Conway.Rules (ConwayBbodyPredFailure (..))
+import Cardano.Ledger.Plutus (SLanguage (..))
+import Cardano.Ledger.SafeHash (originalBytesSize)
+import Cardano.Ledger.Shelley.LedgerState
+import Cardano.Ledger.Shelley.Rules (
+  BbodyEnv (..),
+  LedgerEnv (..),
+  ShelleyBbodyState (..),
+ )
+import Cardano.Ledger.TxIn
+import Control.Monad (forM)
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Map as Map
+import qualified Data.Sequence.Strict as SSeq
+import qualified Data.Set as Set
+import Lens.Micro ((&), (.~), (^.))
+import Lens.Micro.Mtl
+import Test.Cardano.Ledger.Babbage.ImpTest
+import Test.Cardano.Ledger.Imp.Common
+import Test.Cardano.Ledger.Plutus.Examples (alwaysFailsWithDatum)
+
+spec ::
+  forall era.
+  ( AlonzoEraImp era
+  , BabbageEraTxBody era
+  , EraSegWits era
+  , InjectRuleFailure "BBODY" ConwayBbodyPredFailure era
+  ) =>
+  SpecWith (ImpTestState era)
+spec = describe "BBODY" $ do
+  it "BodyRefScriptsSizeTooBig" $ do
+    Just (script :: Script era) <- pure largeScript
+    let scriptSize = originalBytesSize script
+    let maxRefScriptSizePerTx = 200 * 1024
+    let maxRefScriptSizePerBlock = 1024 * 1024
```

Comment:
Why did you decide not to use the top level bindings from the rules?

Reply:
Oh because I wrote this test initially for Babbage, but there's no good reason, will change.